### ISSUE TITLE
Components

### ISSUE THREAD

**Issue by jbucaran** [hyperapp/hyperapp#238]
Let's discuss how components could look like in a future version of HyperApp.

### Submitted Proposals

- https://gist.github.com/zaceno/83286dfd2a18ebbd3b4f1d7cb5810a0c via @zaceno 
- https://gist.github.com/Swizz/8437df6ac5ec4985e72e648d242f6a9f via @Swizz 

---

### `counter.js`
```jsx
const Counter = (props, children) => ({
  state: {
    value: props.value
  },
  view: ({ value }, { inc, dec }) =>
    <div>
      <h1>{state.value}</h1>
      <button onclick={inc}>+</button>
      <button onclick={dec}>-</button>
    </div>,
  actions: {
    inc: state => state + 1,
    dec: state => state - 1
  }
})
```

### `index.js`
```jsx
app({
  state: {
    title: "Hello."
  },
  view: (state, actions) =>
    <main>
      <h1>{state.title}</h1>
      <Counter value={1} />
    </main>
})
```

## Credits
All credit goes to @MatejMazur:
- https://github.com/MatejMazur/hyperapp
- https://github.com/MatejMazur/hyperapp-component-experiment

## Related
- #147 
- #219

**Comment by MatejMazur** [comment]
Yesterday I've created partially (doesn't work without IDs) working solution without `components: []`. Just import component and you can use it but I don't know how to track components without publicly exposed identification. I need some kind of tracking. 

Here is actual and working variant of writing apps with this version:

# Example

## app.js

```javascript
import { h } from 'hyperapp'
import Counter from './counter'
import Logger from './logger'

app({
  state: {
    who: 'world',
    counters: [
      { id: 0, initial: 0 }, // I don't know how to avoid ID
      { id: 1, initial: 1 },
      { id: 2, initial: 5 }
    ]
  },

  actions: {
    addCounter: (state) => {
      const counters = state.counters
      counters.push({ id: counters[counters.length - 1].id + 1 })
      return ({ counters })
    },
    removeCounter: (state, actions, index) => {
      state.counters.splice(index, 1)
      return {
        counters: state.counters
      }
    }
  },

  view: (state, actions) => (
    <div>
      <div>Hello {state.who}!</div>
      <button onclick={actions.addCounter}>Add counter</button>
      {state.counters.map((counter, index) => (
        <div>
          <Counter id={counter.id} initial={counter.initial} />
          <button onclick={() => actions.removeCounter(index)}>Delete</button>
        </div>
      ))}
    </div>
  ),

  mixins: [Logger]
})
```

## counter.js

```javascript
import { h } from 'hyperapp'

export default {
  name: 'Counter',

  state: (props) => ({
    sum: 0,
    number: props.initial || 1
  }),

  actions: {
    increase: (state) => ({
      sum: state.sum + state.number,
    })
  },

  view: (state, actions, props, children) => (
    <div>
      <div>Counter: {state.sum}</div>
      <button onclick={actions.increase}>
        Add {state.number}
      </button>
      {children}
    </div>
  )
}
```

## Gif example

![example02](https://user-images.githubusercontent.com/20944757/27372455-0dbf4148-5665-11e7-9dba-c7b78fc08214.gif)
In the example above you can see app.state with states to every instance under `@Counter: {}`. Every instance is stored under its ID.


# How did I achieve this?

In a nutshell I've achieved this by creating and sharing outerEvents between `app()` and `h()`. When you call `h(tag, data)` it will verify if `tag` is an object and if it is, it calls `outerEvents.get("child", ({ component: tag, data }))` which will get response from `app()` where is `outerEvents.set("child", ({ component, data }) => ...)` and here is look for existing instance by ID in `app.state[componentName][instanceId]` or it creates new one. Actions are almost the same as in the first version (component's actions are connected app.actions with first loading of component). ...It's a little bit less code than in my previous version but I've found out that I can't use components without ID (that problem is even with first version).

# Known issues (TODOs)

- removing instance (I need something like event listener for onremove) 
![issue01](https://user-images.githubusercontent.com/20944757/27378471-44fee2c2-5678-11e7-9b06-38a2e92ec9a7.gif)
- generating instance without ID (maybe we can store ID in DOM attribute and use it for calling instance)
- optimizations (too much code in the core...but I think that you can help me with it if it would be interesting approach)

## Links
[Counter example](https://github.com/MatejMazur/hyperapp-component-experiment/tree/emit-version)
[HyperApp experiment](https://github.com/MatejMazur/hyperapp)

**Comment by Swizz** [comment]
This one looks too magical. For someone like me who is new into the Hyperapp community.

I am pretty in love about the fractal idea for state/actions. The main concern is to handle list of same component with their own state.

**Comment by MatejMazur** [comment]
@Swizz Thanks for your view. 👍 But I think that only magic is ID part (we need some identification for instances of component if we want keep single source of truth but maybe we can hide it somehow). If you look closer. You will see that component Counter (POJO) is "stateful component". So as user you write components almost the same as `app()` (you work with component's state and actions just like in stateful components). Only difference is adding name key (maybe we find some solution without naming) and props with children but they are fundamental for component based approach (you need somehow process inserted properties and children from parent). But these concerns can be easily suppressed in docs.

**Comment by jbucaran** [comment]
@MatejMazur The magical part is what you are doing inside `h` to make this miracle happen. 💯 

The reason it is magical is that custom tags return a virtual node tree, not an object. So, although I like this I also agree with @Swizz.

### ids

I don't think we need to get rid of ids. The id is not really the core of how components work, but rather how index.js keeps track of them. Let's keep thinking. I've updated [my example](https://github.com/hyperapp/hyperapp/issues/238#issue-237408176) to fill in the missing code.

**Comment by jbucaran** [comment]
@MatejMazur How do you pass initial values / props to a component in your example?

**Comment by jamen** [comment]
Awesome @MatejMazur :smile:.  I think a good start to something awesome, looks like a clean solution to me.

Like @jbucaran, I'm wondering how you get more control over the props.  I do see you pass them through `view`, but I think you would need more "open access" over them for all fields, because it is your only method to accessing global state and global actions (by passing them through).

What if you did the exact same thing, but wrapped it in a function that was passed `(props, children)`?:

```js
const Component = (props, children) => ({
  // ...
})
```

Instead of passing them through every single avenue like we do with state/actions.

**Comment by leeoniya** [comment]
> What if you did the exact same thing, but wrapped it in a function that was passed (props, children)?:

FWIW, we recently had the same discussion in domvm [1], since it works this way. the only issue is that if either of those values ever get replaced by a parent, they become stale in the closure.

[1] https://github.com/leeoniya/domvm/issues/147#issuecomment-305987998

**Comment by MatejMazur** [comment]
> @jbucaran How would you pass initial values / props to a component in your example?
```javascript
state: {
  ...,
  counters: [
    { // props are just everything inserted to custom tag
      someValue: 0, 
      someText: 'foo',
      initial: 1 // I want to keep eye specifically on "initial" because everytime you update it, 
                   //  you'll get to the problem with previous state of component...
                   //  what to do in such situation? ..reset state?
                   // I think that it's important to keep such important value separately 
                   // from props inserted only to view part of component which don't 
                   // affect component state but hey..we can change that :-D
    }
  ]
}
...
{state.counters.map((counter, index) => (
  <Counter {...props} /> // variant A
  <Counter 
      value={counter.someValue} 
      text={counter.someText} 
      initial={counter.initial} /> // variant B
))}
```

@jamen

> Like @jbucaran, I'm wondering how you get more control over the props and children. I do see you pass them through view, but I think you would need more "open access" over them for all fields, because it is your only way to get global state and global actions.

Send example of some situation where you need more "open access". Now the `props.initial` is key for more open access.


> What if you did the exact same thing, but wrapped it in a function that was passed (props, children)?

I think that it's important to keep ordinary props separately from component state and actions because if you want somehow to interfere to component actions then why would you do that? You can change behavior of component actions by initial value (it can be str, num, obj, arr) inserted in component state then you can use it through component state in args of action. The same way as you do these things in `app` actions.

**Comment by zaceno** [comment]
FWIW, with one really tiny change in `app.js` + a simple function that can live in userland, it becomes possible to write code like @MatejMazur 's example above. See PR https://github.com/hyperapp/hyperapp/pull/241

... of course this approach breaks the "single source of truth" notion. In my setup each component is it's own app with it's own state. Communication between parents need to happen via props (pass notification actions down to children).

So it's not the perfect solution. (For example, time-travel debugging a bunch of apps that can come and go through the life of an app seems pretty tough, if not impossible) But I thought it relevant to mention in this thread anyway.

**Comment by naugtur** [comment]
Hi everyone. I saw this thread after I already had some expectations as to how components would work and I thought I'd show you what they were.

In a nutshell:
- don't change how components work
- keep one state and one source of actions
- I went with context for passing anything instead of adding a dispatch function; seemed more powerful

See #245

**Comment by jbucaran** [comment]
@naugtur Thanks for the feedback. This thread is to speculate and discuss how the components API could look like. 

> keep one state and one source of actions

Absolutely! 💯

**Comment by Swizz** [comment]
My contribution to the discussion, with a little bit of my Cycle.js thoughts (Onion, Isolate, etc...)
Like I said in favor of Fragmented State and Actions for a Single Source of Truth.

**Here is a working example on Hyperapp 0.9.3** :tada: 

#### `Counter.js`
```js
const Counter = (props, children) => () => ({
  state: {
    Counter: {
      [props.key] : {
        count: 0
      }
    }
  },

  actions: {
    Counter: {
      up: state => ({ Counter: { [props.key] : { count: state.Counter[props.key].count + 1 } } }),
      down: state => ({ Counter: { [props.key] : { count: state.Counter[props.key].count - 1 } } })
    }
  },
	
  view: (state, actions) => 
    h('div', { class: 'counter' }, [
      h('button', { onclick: actions.Counter.up }, '+'),
      h('span', null, state.Counter[props.key].count),
      h('button', { onclick: actions.Counter.down }, '-')
    ])
})
```


#### `index.js`
```js
app({
  view: (state, actions) =>
    h('div', null, [
      Counter({ key: 1337 })().view(state, actions)
    ]),
		
   mixins: [Counter({ key: 1337 })]
})
```

---

As you see, Mixins allow already a lot of things regarding fractal state and actions.
So, I vote in favor of a Component api point : 
```js
app({
 //...
 depends: [Counter({ key: 1337 })]
 //...
})
```

That will be just a super powered mixins that will do the following stuffs :
* Wrap component state into : 
```js
state : Counter: { [props.key] : { STATE } }
```
* Give to a component a framgented state according to Component and Key and wrap the result into the single global state
```js
{ state: Counter[props.key] } => ({ Counter: { [props.key] : { count: state.count - 1 } } })
```
* Give to the view caller and array of component view by Component and `props.key`
```js
view: (state, actions, { Counter }) =>
  h('div', null, [
    Counter({ key: 1337 })
  ]),
```

**Comment by Swizz** [comment]
A little bit more neat with JSX. :+1:
I think I will battletest this on the app I am working on.
(With some helpers for actions to avoid this long-long merge chain)

#### `Counter.js`

```jsx
const Counter = (props, children) => () => ({
  state: {
    Counter: {
      [props.key] : {
        count: 0
      }
    }
  },

  actions: {
    Counter: {
      up: ({Counter: state}) => ({ Counter: merge(state, { [props.key] : merge(state[props.key], { count: state[props.key].count + 1 }) }) }),
      down: ({Counter: state}) => ({ Counter: merge(state, { [props.key] : merge(state[props.key], { count: state[props.key].count - 1 }) }) })
    }
  },
	
  view: ({[props.key] : state}, actions) => (
    <div class={"counter"}>
      <button onclick={actions.up} />
      <span>{state.count}</span>
      <button onclick={actions.down} />
    </div>
  )
})
```

#### `index.js`
```jsx
app({
  view: (state, actions) =>
    h('div', null, [
      (<Counter key={1337}/>).view(state.Counter, actions.Counter)
    ]),
		
   mixins: [<Counter key={1337}/>]
})
```

**Comment by Dohxis** [comment]
I would like to jump into this conversation by saying that I really like where this little framework is going. I have been using Elm for some time now and the best thing for me is that Elm is elegant and simple to use. Its syntactically beautiful. As Hyperapp is trying to be Elm-like I guess seeing those objects nesting one into another just tooks all the beauty. Why we cannot use new Javascript features to have beautiful looking framework. In my opinion, getting things out of those objects and at least letting components have a nice syntax would be a great decision.  I have been working on my app and wrote some helper functions and classes to help me work with it. Here is the simple component which I thing looks simple and easy understand what its doing by just looking at it. 
```javascript
class Counter extends Component {

    constructor(props){
        super(props);
    }

    render(){
        return (
            <div>
                <button onclick={this.actions.add}>+</button>
                {this.state.num}
                <button onclick={this.actions.sub}>-</button>
            </div>
        );
    }
    
};
```
Yes it looks like `React  but its the syntax people have been using for a long time right now, and they like it. Projects always keeps growing and having a nice way to write components will make our lives just easier.

This is only my opinion. Keep up with a great work :+1:

**Comment by naugtur** [comment]
Classes bring side effects and all the mess were avoiding with functional programming. It's all about references to functions and the main structure of the app could be declared once while each component is a pure function. That's what came with elm, classes would hurt that.

**Comment by jbucaran** [comment]
@Dohxis We want single app state architecture. So, using `this` goes against this very principle. :)

**Comment by Dohxis** [comment]
I can understand that classes would bring unwanted things, did not experienced on my small app though. But I am keeping single state and single actions. They are passed as props.

**Comment by Swizz** [comment]
I was thinking about how to implement interactive doc, when I was punch by an awesome thing : 
**Hyperapp is already ready for HTML component by design**

```html
<body>
  <app-counter></app-counter>
  <app-counter></app-counter>
  <app-counter></app-counter>
</body>
```
```js
const { h, app } = hyperapp

const Counter = (props, children, element=document.body) => ({
  root: element,
  state: {
    value: props.value
  },
  view: ({ value }, { inc, dec }) =>
    h('div', { class: 'counter' }, [
      h('button', { onclick: inc }, '+'),
      h('span', null, value),
      h('button', { onclick: dec }, '-')
    ]),
  actions: {
    inc: state => ({ value: state.value + 1 }),
    dec: state => ({ value: state.value - 1 })
  }
})

document.querySelectorAll('app-counter').forEach((elem) => {
  app(Counter({value: 0}, [], elem));
})
```

This is not really great to see ; but it works well : 
```js
app({
  state: {
    title: "Hello."
  },
  view: (state, actions) =>
    h('main', null, [
      h('h1', null, state.title),
      h('app-counter', null)
     ]),
  events: {
    render(state, actions, view, emit) {
      document.querySelectorAll('app-counter').forEach((elem) => {
        app(Counter({value: 0}, [], elem));
      })
    }
  }
})
```

**Comment by jbucaran** [comment]
@Swizz Please keep them coming! It adds a lot to the discussion. 🙏

**Comment by nitin42** [comment]
We could also think about adding `Dynamic Components` using the same mount point and switching between different components by binding them to an attribute like `Vue.js` does.

Example 

```javascript
app({
  state: {
    title: "Hello."
  },
  data: {
    currentView: 'home'
  },
  components: {
    home: { doSomethingHere },
    About: { doSomethingHere },
    Contact: { doSomethingHere } 
  }
})
```    

So depending upon the UI logic we could switch between different component using a wrapper or a method.

This would be useful in doing transitions (transitioning between the components or elements) and other stuff.

**Comment by Swizz** [comment]
@nitin42 Im definitely going to work on this one : https://github.com/hyperapp/hyperapp/issues/238#issuecomment-310999839

This is best one : Single state and actions tree "object" merged with the superpower of mixins. So, all the code base is already here.

**Comment by nitin42** [comment]
Great!

**Comment by jbucaran** [comment]
Very interesting idea @Swizz. Can you give it another shot at explaining the implementation part?

What is `depends`? And how to avoid repetition?

**Comment by Swizz** [comment]
My english is not really good yet, but I will try to be as clear as possible.

`depends` is just another word for `components` but as you seen, if you want two instances of a counter your main app need to depends on two instances of a counter.

I dont know really why, but I am against the fact to declare component into the view.
Components need to be instanciated at high level, view() job is only to make a presentation of the state, isnt it ?

But we need a developer experience when using component by that way will be also possible `app(Component())`. So this is why I am in favor of Fragmented state.

The hyperapp maintain a single source of truth as a single state object with a single object of mutators (actions). 	Through a namespace system. But dynamic because we can handle a lot of instance of a component with their own state substree.

The former example is working with the current implementation with mixins but this is not really neat by using a lot of tricks.

**Comment by MatejMazur** [comment]
@Swizz basically came up with the same approach as I did in Slack channel. But he knows how to achieve that with less intervention to core (congratulation!).
It's not bad, but we want to avoid another boilerplate so we can use the components with the same simplicity as the application itself.

**Comment by jbucaran** [comment]
@MatejMazur @Swizz @jamen @everyone

How does everyone feel about this approach?

## `app.js`
```jsx
app({
  state: {
    title: "Hello."
  },
  view: (state, actions) =>
    <main>
      <h1>{state.title}</h1>
      <Counter value={1} />
    </main>
})
```

## `counter.js`

```jsx
import { h, component } from "hyperapp"

const Counter = component((props, children) => ({
  state: 0,
  view: (state, { up, down }) =>
    <div>
      <h1>{state.value}</h1>
      <button onclick={up}>+</button>
      <button onclick={down}>-</button>
    </div>,
  actions: {
    up: state => state + 1,
    down: state => state - 1
  }
}))
```

**Comment by Swizz** [comment]
I am still against the fact to instantiate components from view function. (but you are the director)

There will be a single state/actions object with subtree/namespace common to the whole app or a component will be a subset of an app capable to live into a main app view ?

How about parent to child communication ? And brother to brother ? Keeping a single state not one by component could facilitate this one.

**Comment by jbucaran** [comment]
> I am still against the fact to instantiate components from view function.

What do you mean by "instantiate components from view function"?


> There will be a single state/actions object with subtree/namespace common to the whole app or a component will be a subset of an app capable to live into a main app view ?

Sub-tree, single state tree.

**Comment by Swizz** [comment]
> What do you mean by "instantiate components from view function"?

```jsx
view: (state, actions) =>
  <main>
    <h1>{state.title}</h1>
    <Counter value={1} />
  </main>
```

Adding a `components` props to the `app()` is IMHO a better pattern.
The `view()` is only to deal with vnode. I think giving it the role to ensure components usage is more than excepted.

A `components` props, could tell clearly "I want to use this component, so add a dynamic segment into the state tree and give me access to its mutators" and component side being an app or a component will be transparent as it will be work on its own mutators and state. As a fractal.

Same thougths than Cyclejs and Onionify but without FPR and more as Elm Architecture.

**Comment by Swizz** [comment]
OR.. We can discourage the use of stateful components. To maintain only a single state as source of truth.

**Comment by jbucaran** [comment]
@Swizz There are no stateful components anywhere.

**Comment by jbucaran** [comment]
> A components props, could tell clearly "I want to use this component, so add a dynamic segment into the state tree and give me access to its mutators" and component side being an app or a component will be transparent as it will be work on its own mutators and state. As a fractal.

Don't like it so much, but I have to agree with you here. 👍

**Comment by MatejMazur** [comment]
So back to this?
```javascript
app({
  state: ...,
  actions: ...,
  view: (state, actions, components) => (...),
  components: [Counter,...],
  mixins: [...]
})
```

**Comment by Swizz** [comment]
Not yet. In my implementation, if you want multiple counters, you will need to use the following : 
```jsx
app({
  components: [Counter(1), Counter(2), ...]
})
```

But how to add another counter from actions ?

Component is just a way to add a subset to an app but actually we need to think about how to interact with a fractal of the state if multiple "instances" of a component exist..

**Comment by MatejMazur** [comment]
@Swizz I solved this before in Slack channel. You don't need init instances in `components`. You need only component and then you use id from array of counters. That will create state under specific id.

**Comment by jbucaran** [comment]
@MatejMazur @Swizz So, I'm thinking like this. 🤔 

### `app.js`

```jsx
import { h, app } from "hyperapp"
import Counter from "./counter"

app({
  state: {
    title: "Hello."
  },
  view: ({ title }, actions) =>
    <main>
      <h1>{state.title}</h1>
      <Counter initialValue={5} />
    </main>,
  components: [Counter]
})
```

### `counter.js`

```jsx
import { h } from "hyperapp"

export default ({ initialValue }, children) => ({
  state: initialValue,
  view: (state, { up, down }) =>
    <div>
      <h1>{state}</h1>
      <button onclick={up}>+</button>
      <button onclick={down}>-</button>
    </div>,
  actions: {
    up: state => state + 1,
    down: state => state - 1
  }
})
```

**Comment by naugtur** [comment]
@jbucaran re: your proposal - I don't like the fact that the component has state that's separate.

That's why in my fork I proposed a version that kept state in the main app and passed down the actions which are bound to the main state anyway.

**Comment by MatejMazur** [comment]
@jbucaran But how do you connect to instance part in global state? You need some kind of id. Because if you don't have id, you can't update state of instance.

**Comment by jbucaran** [comment]
@naugtur The component's state is not separate. It looks _as if_ it was separate, but it is not.

Neither was that case in my first approach. But it was magical, as it wasn't obvious that Counter's state and actions should be merged with the app's intance internal state, actions.

Does that make sense? :)

**Comment by MatejMazur** [comment]
@jbucaran And when you drop the third argument from the view, you actually call the Counter function again without any dependency to Counter in `components` prop. That's weird.

**Comment by jbucaran** [comment]
@MatejMazur Whoops, you are right. We'll need the `component` function after all to wrap Counter.

**Comment by Swizz** [comment]
Component seem to have its own state, but component "state" will be merge to the app state. Like it does for mixins. 
In my mind, component is just a superpowered mixin that just add things into a namespace.

@MatejMazur @jbucaran ?

```jsx
import { h, app } from "hyperapp"
import Counter from "./counter"

app({
  state: {
    title: "Hello."
  },
  view: ({ title }, actions) =>
    <main>
      <h1>{state.title}</h1>
      <Counter key={1234} />
    </main>,
  components: state => [Counter({ initialValue: 5, key: 1234 })]
})
```

```jsx
import { h, app } from "hyperapp"
import Counter from "./counter"

app({
  state: {
    title: "Hello."
  },
  view: ({ title }, actions, components) =>
    <main>
      <h1>{state.title}</h1>
      {components.map((Component, /* key */) =>
        <Component /* key={key} */ />
      )}
    </main>,
  components: state => [Counter({ initialValue: 5, key: 1234 })]
})
```

**Comment by MatejMazur** [comment]
@jbucaran And we're back again. 😄

**Comment by jbucaran** [comment]
@MatejMazur 🤣

**Comment by naugtur** [comment]
How about this: 
(should be doable based on my experience to date)

**app.js**

import { h, app } from "hyperapp"
import Counter from "./counter"

app({
  state: {
    title: "Hello.",
    counterItem: 5
  },
  view: ({ title }, actions) =>
    <main>
      <h1>{state.title}</h1>
      <Counter />
    </main>,
  components: [Counter]
})

**counter.js**

```js
import { h } from "hyperapp"

export default (props, children) => ({
  view: (state, { up, down }) =>
    <div>
      <h1>{state}</h1>
      <button onclick={up}>+</button>
      <button onclick={down}>-</button>
    </div>,
  stateRoot: "counterItem",
  actions: {
    up: state => state + 1,
    down: state => state - 1
  }
})

```

Names of things are not part of the suggestion, stateRoot is a lame name, but easy to understand the concept here.

I'm not sure if the decision about state root should be made inside or outside the component.

**Comment by MatejMazur** [comment]
@naugtur could you please correct format of code?

**Comment by naugtur** [comment]
I did right away, but github doesn't refresh automatically if it's not a new post

**Comment by jbucaran** [comment]
This one is tough, but we'll make it through. I am sure. I've noticed a few more mistakes above, so I'll take another approach to brainstorming this by creating a list of facts.

- Components could be either a function or an object, but its definition must be wrapped in a helper function like `component(...)` or the component **must** be passed into the view. This is necessary because `<Counter initialValue={0} />` is equivalent to `h(Counter, { .. })`. 
- `app()` can be called multiple times in a single application. So, we need a way to register components with each specific app instance's internal state/action tree.
- Component state/action tree should be merged with the app state/action tree during app start, not later during render. This is because some components may be rendered conditionally, so there's no way to reliably know which vnodes correspond to components without changing how `h` or `patch` works and we shouldn't since those bits are off limits.

**Comment by naugtur** [comment]
I'd add to this a suggestion to define state in one place, outside components. That's what redux got its adoption from and imho it makes it easier to explain or understand the app.

If I can push it even further, I'd suggest to only have components as pure functions that get a reference to a part of state and some actions. That'd eliminate the need to merge anything on app start and would make component scope (by scope I mean part of state they access) independent of component parent-child relationship.

**Comment by jbucaran** [comment]
@naugtur Can you post an example with fake code? 🤔

**Comment by naugtur** [comment]
You mean for the first paragraph or second?

**Comment by nitin42** [comment]
@Swizz One thing !

Let's say we are switching-out the components but what about their state when they are switched ? 
We should preserve their state in memory so as to avoid re-rendering (performance issue). Right ? 
This way we could cache or store the components which are not active.

**Comment by naugtur** [comment]
Passing a reference to app to components is still an unknown to me, but merging states and actions seems even harder. Anyway,

counter/Counter.js
```js
import { h } from "hyperapp"

// this part could be wrapped in a function call, but I don't see a reason to
export default (props, children, actions) => (
    <div>
      <h1>{props.state}</h1>
      <button onclick={actions.up}>+</button>
      <button onclick={actions.down}>-</button>
    </div>
)
```

counter/actions.js
```js
export default {
    up: state => state + 1,
    down: state => state - 1
  }
```

```js
import { h, app, scopedActions } from "hyperapp"
import actions1 from "counter/actions.js"
import Counter from "counter/Counter.js"

app({
  state: {
    title: "Hello.",
    counterItem: 5
  },
  view: (state, actions) =>
    <main>
      <h1>{state.title}</h1>
      <Counter state={state.counterItem} />
    </main>,
  actions: Object.assign({},  scopedActions("counterItem",actions1), ...)
})
```

That scopedActions part I'm not proud of, but the goal is to let more than one component share an action through an obvious reference.

If we put actions in components, how do you prevent them from name collisions while keeping them available to other components? And don't tell me they shouldn't be available - all redux projects I've seen had reducers in multiple features listen to a shared action type (eg. rerendering something when you close a modal)

**Comment by Dohxis** [comment]
As my first suggestion was bad (I realized why), I was experimenting more and wrote everything as an objects 
```javascript 
const Counter = {
    view: (state, actions) => (
        <div>
            <button onclick={actions.add}>+</button>
            {state.number}
            <button onclick={actions.sub}>-</button>
        </div>
    ),
    actions: {
        add: (state) => state.number + 1,
        sub: (state) => state.number - 1,
    }
};
```
I really like how simple it looks and how easy it would be to add features to it. What do others think?
My suggestions are:
* Let define component's own actions, its easier to think about a component if it comes with its own actions. 
* Pass a state for all components so we would have a single source of truth.

**Comment by naugtur** [comment]
@Dohxis your current suggestion repeats a lot of bits already printed above, but it doesn't explain how state would be scoped or actions merged. If you ignore these two aspects, your suggestion is practically the same as my previous one.

**Comment by Dohxis** [comment]
@naugtur Yes, I just want to support the suggested component design. No wrappers like `component()` simple plain objects. Talking about actions I guess it would be nice if all actions from all components would be merged into one under its own namespace. I am sure there is a better way to do this but having functionality like this would help I think.
```javascript
const Counter = {
    view: (state, actions, globalActions) => (
        <div>
            <button onclick={actions.add}>+</button>
            {state.number}
            <button onclick={actions.sub}>-</button>
            <button onclick={globalActions.OtherComponent.doSomething}>Go wild!</button>
        </div>
    ),
    actions: {
        add: state => ({ state.number + 1 }),
        sub: state => ({ state.number - 1 }),
    }
};
```

**Comment by andyrj** [comment]
@naugtur I'm a little worried there still needs to be some key/id for the components even if we have scopedActions(), or people are going to be mucking up the state thinking that it's only in the component or did I miss a part there?

What if you made the Component() a part of the hyperapp api and had it extend the state and generate id's for each component?

NOTE: very rough not working code just trying to express my thought
```js
function Component(name, {state, views, actions}) {
  let id = generateId();
  /* I'm thinking components extend the global state atom something like 
     {
        Components: {
           name: {
             id: {
                // state specified by component goes here...
             }
           }
        }
     }
  */

  // and they can register name, to a map for components {state, view, actions}
  // maybe it could wrap actions, kind of like scopedActions in naugtur's example above.
  Object.keys(actions).forEach((a) => {
    return (globalState) => a(globalState.Components[name].id) // I'm thinking actions would need to be curried or partially applied
  })

  //  then as long as the actions returned state relative to the described state given to Component(), we would know where to merge it...
  // down side would be then it would have trouble changing state outside of it's component scope, or we would need to create another way to differentiate a scopedAction and non scoped action...

  // return some object that will be useful on render passes to look up component by name and id?
}
```
Then the part of the code that checks for tagName or function in h.js, could be modified for components by name instead of just using the function as is?

I don't know that might be to far off from the elm way, been looking at the docs for scaling the elm architecture and I really don't like elm syntax lol.

**Comment by MatejMazur** [comment]
You just rewrited my first idea. 😄 It begins look messy here..there are a lot of similar ideas. Problem with my approach is in third point of @jbucaran [comment](https://github.com/hyperapp/hyperapp/issues/238#issuecomment-311743386).

**Comment by zaceno** [comment]
I'm trying to jump into this topic, trying to follow along on the progress so far. Quite bewildering but fun :) Anyhow, one question stands out for me:

(please excuse me if this has already been talked about, and please nobody take this as any kind of criticism -- it's a well-meaning and honest question)

It is very clear we are aiming to keep all the state in a single state-tree. I'm all for that. As I understand it, the main reason for having a central store for all state (and the main motivation for libraries like Flux, Redux, Vuex...) is because apps with only locally-stateful-components end up with very messy, hard to follow "cross-talk" -- because they need to share a lot more state than initially anticipated.

But the patterns I've seen so far have state and actions that are defined very independently. Even though we're merging their state and actions to the main tree, they don't seem to have any access to the main state tree. So how are the components meant to communicate? If the answer is: "via the props", then they are effectually no different to locally stateful components. And if that's the case, then what's the point of the "one-state-to-rule-them-all"?

So I guess what I'm saying is: it looks to me like were trying to emulate locally-stateful components, and I don't see how keeping their state in the main tree will save us from the problems that locally-stateful components exhibit in frameworks that support them.

So, clearly there's something here I'm not getting ;) Anyone care to fill me in?

**Comment by andyrj** [comment]
@zaceno I agree that it will be emulating stateful components but there could be some way of escaping the scoping, maybe extend actions to have global and local.  I understand the concern, but also see where it could be implemented in a way that makes some applications easier to express.

**Comment by zaceno** [comment]
@andyrj I agree that locally-stateful components, judiciously used (wether emulated or not) can make an application easier to express. I just feel like if (mostly) locally scoped components are what you want then just let the scope be local, don't bother with merging it to the main state tree. Do what Vue, React, et al do and let the components communicate via props and/or event bus. 

So I'm not opposed to locally scoped components (on the contrary, I use them) BUT *that can be solved outside hyperapp*. There's no reason for hyperapp to explicitly support local-state components.

**Comment by andyrj** [comment]
ya I looked at your example and was impressed by the dark arts ;) (referring to @jbucaran's comment on it)  I didn't fully grasp how it worked though I am gonna have to look over it again.

**Comment by zaceno** [comment]
@andyrj: Lol, yes I'm known to break a taboo or two :) Actually don't look too closely at that example, because I realized it doesn't update when you pass new props to it. So there's a little more work to do with it, but I'm certain it can be made to work. The basic notion is that a locally scoped component is just it's own `app(...)` -- you just need to solve how and where to mount it.

**Comment by zaceno** [comment]
Not to be too "down" on this thread, I must say: I see that there are a few real and important problems we're trying to solve here (just not sure if components are the answer). And those are

- How we can scope / fragment state. So actions & views can be reused for different parts of the state tree without being dependent on the structure of it. 

- How we can work with a dynamic state tree. Instantiating new "branches", reordering them et c (in conjuction with having bound views and actions)

At least the first problem, I've seen some nice patterns presented on slack and elsewhere I think. Perhaps that's enough until the problem is more clearly understood?

**Comment by naugtur** [comment]
@zaceno In my examples (including the first one in repo that actually works) I focus on having one state that's bound to all actions no matter where they are. Latest example is also scoping the state so that each action is not forced to be able to handle the whole state object (like combining reducers in redux). 
Components can only interact by:
- passing properties down
- calling actions from other components (effectively reloading a parent list etc.)

I'm not a contributor here (not saying I won't become one if opportunity presents itself) so feel free to disregard my opinion, but I think local state of a component that is a pure function is as easy as adding a closure and can successfully be ignored in this discussion.

**Comment by zaceno** [comment]
@naugtur Your opinion is every bit as valid as mine! :) Anyway I'm not trying to present an opinion so much as understand. Clearly you guys in here have thought a lot about this.

So if the type of components you're describing have access to the entire state and set of actions (or all the actions at least) -- that certainly is something different from what I was percieving. Definitely not "emulated local scope". I think there's probably some value in that approach!

(And yes, as I said before, true local scope can be handled outside hyperapp and we can disregard that for this discussion. I only brought it up because it looked like we were on the road to that...)

**Comment by zaceno** [comment]
So with the risk of causing even more confusion (and possibly repeating what someone in here has already proposed), here's my thing to add to the mix:

https://codepen.io/zaceno/pen/gRvWPx?editors=0010

No magic whatsoever with the state & actions (I'd prefer to leave that to some other tool, like lenses which @lukejacksonn has suggested). In this concept, components are nothing more than view functions bound to the same state and actions as the main view function. Made available through a parameter to both the view and other components. It's basically just syntactic sugar so you don't have to pass the state and actions to every sub-component that needs them

**Comment by andyrj** [comment]
@zaceno the only thing I don't like about that example is how the main app has to know all the state needed for the components.  

I kind of feel like the benefit of components would be allowing that to be delegated while still storing it in the single state atom.

I also think what you mentioned in the other issue about returning emit from app() could be a really good place to look at making a component() that can be external to the main app() so that it doesn't require changes to the current code.

**Comment by zaceno** [comment]
@andyrj : right, that's kind of the point I was making before. All of it except the part about still storing it in the single state atom. If you want your components state to be independent and unknown to the main state -- why even store it there? The way I see it, if that's what you want, then go for a external/locally stateful component.  (and you're right, the app-returns-emit PR could help with that too)

Hence my example explicitly avoids doing anything with the state. All state available to everyone all the time :)

BTW, if you're interested in separating state & actions into different modules, you could use mixins.

**Comment by Swizz** [comment]
I didn't read all the comments since at least two days, but I think we are on the way to discuss the concept even if concept is clear in @jbucaran mind. https://github.com/hyperapp/hyperapp/issues/238#issuecomment-311743386

The most relevant purpose is to discuss how to achieve that, isnt it ?
Everybody seem to be sticking to his guns, so we need to reach a consensus.

**Comment by zaceno** [comment]
@Swizz You're right there seems to be little consensus on what components  should be and what they do. I think what we need is a good example app that could be rewritten with the various concepts to demonstrate their benefits. It's hard to reach consensus if everyone is thinking in the abstract, and has their own hypothetical use-cases in mind.

The question is, what would such an example app be? We're all writing counter-examples, but counters are too simple to show the benefits of components.

As for @jbucarans comment, I'm still wondering about this point:

> Components are not stateful. Whatever they expose: state, actions, events should be merged back into the mothership state/action/events tree 

I understand, and basically agree with the sentiment, but:

a) EITHER components are given restricted / no access to the rest of the system state. In which case they are *essentially locally scoped* for all practical purposes, and there's no point in merging the state back.

b) OR components are given full access to the mothership state / access. In that case, why have *multiple references to the same data* inside components? Why not manage all the state in the mothership in the first place?

So I can't really see a scenario where components (as in "sub-apps") managing their own state makes sense.

I do see a point to having helpers (like lenses) to access & manipulate state in a scoped fashion. But I think that is something we solve separate from components.

**Comment by jbucaran** [comment]
Can anyone help summarize the different proposals? I don't want to get bog down in implementation details, but just touch on the surface API (which is all I wanted to discuss in this issue). 

What about a gist for each proposal and create a list of them here?

**Comment by zaceno** [comment]
@jbucaran: good plan. Just documenting API's, right? -- no lengthy example apps?

**Comment by zaceno** [comment]
Component API proposal: "Bound view functions"
https://gist.github.com/zaceno/83286dfd2a18ebbd3b4f1d7cb5810a0c

**Comment by MatejMazur** [comment]
I must laugh when I see how everyone uses the word "magical", but it's only magical because of missing documentation. 
However the goal of the components is not to complicate the code, but to facilitate work independent of the main structure. Zaceno's example is not easy readable. That doesn't solve the issue neither generally proposed `components` prop/arg in `app()` because imagine that you need ten or twenty components.. This quickly becomes unreadable and full of repetitions (you must write component name multiple times). So I don't see any really useful solution yet (neither mine because of problem with multiple apps).
Single state of truth was goal because of better history maintainence and time travel debugging and easy implementation of HMR. Maybe it's not useful to everyone but if we use multiple states then we create another similar library as React, Preact etc.. and only difference will be in app structure. Maybe it's enough.. I'd like create apps without classes.

**Comment by jbucaran** [comment]
@MatejMazur No need to waste energy discussing stateful components. We're not going to implement them. Ever.

@zaceno Thanks for your proposal. I've updated the top issue to add it. 🎉

**Comment by MatejMazur** [comment]
Then this discuss is mainly about how to initiliaze components, how to create instance state and connect it to app state and how to keep simple writing of code without a lot of repetitions.
Example with counters is good enough because imagine how you must write example with multiple counters where you can add or remove one of them. Then you realize that this in Elm architecture needs to know even the structure above it but that's not its interest.
Connecting component actions is relatively easy. But there is huge problem with calling components because you need to know which app called component and you need id of instance and this is where all problems begin.

**Comment by Swizz** [comment]
@jbucaran Is there a due date ?

**Comment by jbucaran** [comment]
@zaceno Wow.

**Comment by jbucaran** [comment]
@zaceno Did you just basically single-handled figured it all out?

**Comment by zaceno** [comment]
@MatejMazur 
> Don't worry how it will complicate structure of HyperApp core. Firstly we need to know if it's even possible to achieve something similar to this.

Yes that's already possible (or close enough anyway) with the current version of hyperapp. But *not* if you plan to merge the state back into the main state. You run into troubles of keeping track of multiple instances, just like you said.

Additionally I'm arguing that there's no clear value in trying to write your app that way AND ALSO keep all your state in the mothership *anyway*. So if that's how you want to write your code (and I wouldn't blame you) there's no reason to wait for anything in hyperapp. (Except you may need https://github.com/hyperapp/hyperapp/pull/259)

**Comment by zaceno** [comment]
@jbucaran , ...um I not sure 😜 I'm hardly sure I know what we're trying to figure out

**Comment by jbucaran** [comment]
@MatejMazur What's wrong with @zaceno's solution?

**Comment by MatejMazur** [comment]
@jbucaran view: `components` arg and prop...you must insert mutliple times a component name...once in import, then in component and then in view args and then (finally) in view returned value.

**Comment by jbucaran** [comment]
For example, `AddCounterButton` shows up three times:

```jsx
    AddCounterButton: (state, actions) => (
      <button onclick={actions.addCounter}>Add counter</button>
    )
  },

  view: (state, actions, {Counter, Total, AddCounterButton}) =>
    <div>
      <AddCounterButton />
      {state.counters.map((v, i) =>
        <Counter index={i} />
      )}
      <Total />
    </div>
```

**Comment by zaceno** [comment]
the destructuring in the view is just so I can use it with jsx. could just do:
```
view:  (state, actions, components) =>
 {components.AddCounterButton()}
 ...

```

**Comment by jbucaran** [comment]
@MatejMazur The following signature:

```jsx
view: (state, actions, {Counter, Total, AddCounterButton}) =>
```

is the same one you originally proposed. I didn't like it then, because I thought this could be implemented in a different way, but just like mixins and actions, we need a way for `app()` to inject them with code.

**Comment by MatejMazur** [comment]
@jbucaran Yes, I know but once you've told me that it will begin full of components then it doesn't make sense to achieve that.

**Comment by zaceno** [comment]
... uh oh -- Did I just bust in here at the last minute and push everything back to square 1? Hope not!

**Comment by jbucaran** [comment]
@MatejMazur Can you try rephrasing that sentence? I think I lost you.

@zaceno Veni, vidi, vici. 🎉  

This looks pretty good, but I still have one issue with this in that, what if you are using 20 components?

**Comment by MatejMazur** [comment]
@jbucaran Sry for my English. But you wrote it. Imagine 20 or more components.. that's undesirable.

**Comment by jbucaran** [comment]
@MatejMazur Ah, is that what you meant? Okay, yeah, I agree on that. This would be fixed if components could have their own sub components. No?

**Comment by zaceno** [comment]
@jbucaran @MatejMazur  I agree the destructuring of 20 components in the view function would not be pretty. But probably also not necessary. In my proposal, every component also gets access to the components-object. So you only need to destructure out the handful you need in each function

**Comment by MatejMazur** [comment]
@jbucaran maybe...but it will increase boilerplate (multiple times component name).. I don't like it. Why did you change your mind on that? 😄 You didn't want that. 😄 Now we've switched opinions.

**Comment by zaceno** [comment]
Still gotta say though -- I'm not that big a fan of my own proposal ;) Not that I see problems with it. Mostly that I don't see it adding a lot of value

**Comment by MatejMazur** [comment]
@zaceno Now I don't either. 😄

**Comment by zaceno** [comment]
@MatejMazur :) 

But about the increased boilerplate: It's the exact same situation with actions. Right?

**Comment by jbucaran** [comment]
> @MatejMazur Now we've switched opinions.

It's not a matter of opinion, but I don't think we have other choices. @zaceno's proposal is at the surface level, pretty much what you had before, almost.

> @zaceno ... uh oh -- Did I just bust in here at the last minute and push everything back to square 1? Hope not!

No, not at all. We're not back to square one. 😄  

For starters, your API design is a bit different to what @MatejMazur originally had (not for a lot though, but your use of props instead of an array makes a huge difference). 

We also have come a long way (I have and I am sure so has @MatejMazur) studying many of the possibilities. 🤓  

We also have a tiny implementation which you've provided via that CodePen.

**Comment by jbucaran** [comment]
@MatejMazur Remember that components are going to live in their own file.

**Comment by jbucaran** [comment]
@zaceno We need components to be able to have their own components. Do you have any idea how we could achieve that with what you have?

Also, maybe you can open a PR? 🔥

**Comment by MatejMazur** [comment]
@jbucaran Yeah, I know that but this is not problem. 😄 Problem was with `components` everywhere (as prop in app, as argument in view, without destructuring you need write longer version of arrow function with return statement (assigning components to variables) or you don't use JSX for components, and it's not usable for more than 5 or 10 components). And even if you nest components it still doesn't look cleaner.

**Comment by zaceno** [comment]
@jbucaran Sure I'll put a PR together. And perahaps a demo of components using other components. And mixins with components (a k a Widgets :) ) It's good to have concrete examples to discuss :)

**Comment by Swizz** [comment]
Single state tree, single actions tree and single component tree.
With a system of namespaces, namespaces for component name and namespaces for isolation ID like it does in Cyclejs. 

Everything live into the same objects at start, but is pretty isolated. When the view function of component is called, we give to it a substet of the tree according to its isolation.

Like a Fractal. Because my english is poor : 
![Fractal state/actions explanation](http://i.imgur.com/7qn3xvG.png)

So, we can create component into seperate file, but when a component as added to an app, it will be merged in isolation.

This will allow to use componenet as app too : 
```jsx
app(Counter(props));
```

**Comment by MatejMazur** [comment]
@Swizz Yes! That's exactly what I want to achieve! That's how I did it but I don't know how to store and calls instances by id if it's allowed to have multiple apps.

**Comment by Swizz** [comment]
I think this is not too dificult. When a component is registered, we can give it a random ID, if one is not given (like key for vnode).

I did something similar with mixins and ES6 features https://github.com/hyperapp/hyperapp/issues/238#issuecomment-310999839

**Comment by MatejMazur** [comment]
@Swizz Yep, but again, you just need another prop in app. So we're back at the beginning. I achieved this without another prop and arg but I don't know how to keep communication between the right app and component - so you need id for app but then you need to tell to component which app calls it.

**Comment by zaceno** [comment]
@Swizz @MatejMazur just don't forget about the need to access the rest of the state and actions. A counter is a very unusual example, in that it's own state/actions are enough for it.

I'd be (honestly!) interested in seeing an implementation of an example where the components as well as the main app need to access eachother's state and actions.

**Comment by Swizz** [comment]
Component are stateless, so, we do not have component which are part of two apps. Here is the same "Model" but here is two different component.

As the Component and the App live with the same State and Actions tree, this is easy to handle parent->child comunication, the Parent could call every action of its childs. 
And for child->parent, this is as easy as adding a `.parent` into the partial component state.

**Comment by MatejMazur** [comment]
@zaceno This isn't difficult because you can set which part of state would be accesible. So we can create for example something like:
```js
component(state, actions, props, children) => {... 
// call app state
state.global.something
// call component state
state.something
}
```
You can just set state before calling component.

**Comment by MatejMazur** [comment]
@Swizz But do you think that it's not possible without some dependency like you have in [comment](https://github.com/hyperapp/hyperapp/issues/238#issuecomment-310999839), don't you? Because it increases boilerplate..

**Comment by zaceno** [comment]
@Swizz & @MatejMazur Yeh, it's possible sure. Still I feel like in a more typical example with lots of cross-component-state-sharing, it wont look as clean and nice as the counter examples. In fact I suspect that *not* using fragmented state will look *cleaner* in those more typical cases. But I'd love to be proven wrong!

**Comment by MatejMazur** [comment]
@zaceno It it's possible than it will be much more cleaner because you don't need to know full path to instance's state, actions etc. You just simply build component like [this](https://gist.github.com/MatejMazur/440bdc8a3ca33604cb0cf1d5d866cdd8).

**Comment by Swizz** [comment]
We can rearrange the API to be more user friendly. And in fact, accessing parent state/actions is a terrible pattern. 
In most case, this is more comfortable to use properties for `parent->child` communication and events for `child->parent` comunication

```jsx
import { h, app } from "hyperapp"
import Counter from "./counter"

app({
  state: {
    title: "Distance",
    unit: "km",
  },
  view: (state, actions, components, emit) =>
    <main>
      <h1>{state.title}</h1>
      {components.Counters.map((Component) =>
        <Component/>
      )}
    </main>,
  components: state => ({
    Counters: [
     <Counter initialValue={5}, unit={state.unit}, onchange={value => emit('changed', value)}/>
    ]
  )},
  events: {
    changed: (state, actions, value) => { alert(value) }
  }
})
```

**Comment by zaceno** [comment]
@MatejMazur Yeah I know but:
a) Your example is still using counters, which do not demonstrate typical complexity.
b) You can use assocPath to  bind/scope actions & state to specific state-branches. Why bake it into a component solution and make it non-optional?

**Comment by MatejMazur** [comment]
@Swizz Agree with parent state/actions. I just want to argue that it's possible. 😉

**Comment by Swizz** [comment]
@MatejMazur We can add a parent propertie to each substract of state and actions tree for those who like to deal with terrible patterns.

@zaceno Do you have a more complex example in mind ?

**Comment by MatejMazur** [comment]
@zaceno You can use assocPath and that's how I do but it can be done automatically. If you build app you don't want work with something what is not related to business logic. And there is still possibility to use assocPath but you still need to know so many things about parent and that's antipattern. Component based approach is about ignorance of parents. 😄

**Comment by MatejMazur** [comment]
@Swizz My solution was based on outer event mediator which sends instance's state, actions to component there it connects these things together then it returns new instance's state to app for storing. But this mediator didn't know id of app so if you want multiple apps then I don't know how to achieve that.

**Comment by MatejMazur** [comment]
@Swizz We can easily create example by using any existing project and our goal would be just insert this project as component to another project (e.g. insert tic tac toe to website built with HyperApp).

**Comment by zaceno** [comment]
@MatejMazur It's no secret that *I like* the component based approach. I like not having to know about stuff 😜 , 

Still don't think hyperapp should to do anything special with states/actions. In my world-view, the counter-examples could be their own apps.

I like @Swizz idea of passing the emit function to views. @jbucaran suggested that too at some point, but can't remember the context.

**Comment by zaceno** [comment]
@Swizz More typical example imo: Anything with forms and validation. There could be multiple types of inputs (sliders, calendar widgets, autocompleting text inputs) and they would each have validation rules based on what was in the other fields. (Like if you selected a weekend in the calendar app, you could only select a certain range on a time slider, stuff like that)

**Comment by MatejMazur** [comment]
@zaceno everything related to communication between components is based on storing props in their parent so this is not purpose of component but rather philosophy behind one way data binding.

**Comment by zaceno** [comment]
:) I think I should just shut up and let @MatejMazur @Swizz @naugtur and @andyrj settle this...

**Comment by MatejMazur** [comment]
Imagine reverse React+Redux approach. Stateful components in React are large apps in HyperApp (multiple apps) and Redux/Elm approach is similar to behavior in these apps. App knows everything inside even local states of components (keep in mind that these states are stored in app.state). So you use mutiple apps only for really large and totally independent parts and components will be used for building these apps.

**Comment by Swizz** [comment]
@zaceno In your example, in my mind, Disabling an input is the job of the parent not the input itself.

I am comming from Vue an Cyclejs, and from both, Component isolation is a panacea.
Component is not allowed to think outer their own business. 

You are attempting to give to more power to component. That why we are alot against Hyperapp capabilities. Because we are not aligned with Elm principle.

A component is a function. And an app is a component which is the root of the component tree.

**Comment by Swizz** [comment]
> Stateful components in React are large apps in HyperApp (multiple apps)

That why, I am moving from the Hyperapp router to a global one to settle an Hyper'app per route.

**Comment by zaceno** [comment]
@Swizz 
> And an app is a component which is the root of the component tree.

All I'm saying is: if you want to structure apps that way with hyperapp, then an app is a tree of apps. If you also want a common state-store for all your components/apps, like Vuex, then that's another app (your store app).  That's fine with me.

Doesn't mean we need to change anything about how `app()` works (maybe there are some minor things, but not sure what yet)

**Comment by MatejMazur** [comment]
@Swizz could you join to Slack channel? Maybe we can try to build something or maybe we don't solve anything. 😄

**Comment by Swizz** [comment]
@zaceno 

Here, we took the decision to maintain only one store (Vuex in your example) and give to each component a subset of this store according to its isolation. We have only one State and Mutators tree by app.

A component is just a way to add features to your app. But it is not either a completely independent feature. When we are talking about component, we are just talking about a Blueprint to add to the app to give it more functionnalities, like a mixins but that could be composable and mutlipliable.

Thats why, I immediately jumped into the mixins logic to present my solution.
Because components are super-powered isolated replicable mixins.

---

@MatejMazur I cant use Slack at work, yeah Slack is for work team, but my IT department consider it as "distraction".

**Comment by jbucaran** [comment]
@Swizz Let's go back to [this](https://github.com/hyperapp/hyperapp/issues/238#issuecomment-311288539) comment for a second.

The problem I have with that is the Counter's actions:

```js
  actions: {
    Counter: {
      up: ({Counter: state}) => ({ Counter: merge(state, { [props.key] : merge(state[props.key], { count: state[props.key].count + 1 }) }) }),
      down: ({Counter: state}) => ({ Counter: merge(state, { [props.key] : merge(state[props.key], { count: state[props.key].count - 1 }) }) })
    }
  },
```

**Comment by Swizz** [comment]
@jbucaran This is because, with this naive implementation, it is possible to give fractal state but not to work with. This need to be done by the core when merging partial state to the main state tree.

**Comment by jbucaran** [comment]
@Swizz Can you write the example again using pseudo-code and call this your proposal? 🙏

**Comment by Swizz** [comment]
Here we go : https://gist.github.com/Swizz/8437df6ac5ec4985e72e648d242f6a9f

This is a complexe but complete example. (no components list yet)

With component props initialisation, `parent->child` and `child->parent` communication, Fractal state, and namespacing.

**Comment by andyrj** [comment]
@zaceno  I definitely don't want to take over anything, I was only chiming in with my 2 cents, I think all the feedback in this thread is useful.

@Swizz I like that gist, but I still feel like we should have some kind of helper or pattern to handle the key/id and at least a pattern for how the state could be partitioned for components.  Maybe we even make the helper generic enough that it can allow users to describe their own lookup() for a components scoped state?

```js
/* example state layout
{
  Component: {
    name: {
      id: { 
        //local state here
      }
    }
  }
}

const stateHelper = (state, name, id) => (fn) => {
  return fn(state, name, id);
}

const exampleThunk = (state, name, id) => {
  return state.Component[name][id];
};
```

Is that kind of the idea of the lens stuff people have been mentioning above?

**Comment by jbucaran** [comment]
@Swizz Why are you using `emit`? 🤔 😕

**Comment by Swizz** [comment]
@andyrj This will allow components to use the state out of their isolation. And I think this is not a good pattern.

@jbucaran `child->parent` comunication. The `change` event call the function on the `onchange` attribute.

We can also move into a `parent` property on `state` and `actions`, if you are more comfortable on this than props/events.

**Comment by jbucaran** [comment]
@Swizz Using emit like is only optional right?

**Comment by Swizz** [comment]
@jbucaran Only useful if you want to use the events system to interact with your parent

**Comment by jbucaran** [comment]
@Swizz I personally find the way component is declared and then used is unintuitive. The declaration should be no different to how we declare actions or events.

**Comment by Swizz** [comment]
I use JSX because the API allow it. But, you can simply use function : 

```jsx
app({
  components: (state, actions, emit) => ({
    Counter: Counter({ 
      initialValue: 5,
      unit: {state.unit},
      onchange: value => emit('change', value)
    })
  })
})
```

or if you do not need anything from the state, actions, emit : 
```jsx
app({
  components: {
    Counter: Counter({ initialValue: 5, unit: 'monkeys' })
  }
})
```

**Comment by jbucaran** [comment]
@Swizz But why do you have to initialize the component when declaring it?

**Comment by jbucaran** [comment]
@Swizz So, why not:

```js
  ...
  components: { Counter },
  ...
```

**Comment by Swizz** [comment]
Because I dont want either to do it into view. Like I said before, Components are in my mind super-powered isolated replicable mixins. Some bluprint with add features to the main app.

So, this initialisation is more than what to add to the State to initialize it. And also a way to isolate Component and add the ability to deal with an array of same components.

**Comment by jbucaran** [comment]
@Swizz So how do you pass props down to the component in the view?

**Comment by Swizz** [comment]
Why do you want to pass props into the view ? 

In this idea, the view role is only to receive the component view vnode result and add it into its children array.
The component hook is the only one responsible do deal with component logic.

View is to make a presentation of the state and add some hooks to deal with actions : no more.

**Comment by jbucaran** [comment]
> Why do you want to pass props into the view ?

I think is is more intuitive to do it in the view. Let me see harder 🔎 🤓.

> The component hook is the only one responsible do deal with component logic.

Sorry, what is the component hook?

**Comment by Swizz** [comment]
I dont know how to call these things ^^
```jsx
app({
  components: ...
})
```

> I think is is more intuitive to do it in the view

You, are the boss. But I am still against :boxing_glove: 😄

**Comment by jbucaran** [comment]
They are called properties.

The following code looks like when you instantiate a component in React. 

```jsx
  components: state => ({
    Counter: (
      <Counter initialValue={5}, unit={state.unit}, onchange={value => emit('change', value)}/>
    )
  }),
```

It's also not clear to me when is this function called and how many times is it called. I guess only once so it's intializing the view?

**Comment by Swizz** [comment]
I am not clear about that, in fact. Initialisation maybe. But what about conditional component or rendering a list of components according to an HTTP request result ?

**Comment by jbucaran** [comment]
@Swizz What about them? I don't see an issue with that because the wrapped component view is passed to the application view in the function like:

```js
  view: (state, actions, { Counter, ... }) => ...
```

**Comment by jbucaran** [comment]
@Swizz What about this?

```jsx

const Counter = (props = { initialValue: 0, unit: "" }, children) => ({
  state: {
    count: props.initialValue,
    unit: props.unit
  },

  actions: {
    inc: (state, actions, data, emit) =>
      emit("change", { count: state.count + 1 }),
    dec: (state, actions, data, emit) =>
      emit("change", { count: state.count + 1 })
  },

  view: (state, actions) =>
    <div class={"counter"}>
      <button onclick={actions.inc}> + </button>
      <span>{state.count} {state.unit}</span>
      <button onclick={actions.dec}> - </button>
    </div>
})

app({
  state: {
    title: "Best counter ever",
    unit: "monkeys"
  },

  view: (state, actions, { Counter }, emit) =>
    <main>
      <h1>{state.title}</h1>
      <Counter
        initialValue={5}
        unit={state.unit}
        onchange={value => emit("change", value)}
      />
    </main>,

  components: { Counter },

  events: {
    change: (state, actions, value) => {
      alert(value)
    }
  }
})
```

**Comment by Swizz** [comment]
How can I use 3 counters  ? An a unknown numerous of counter based on an initialValues array ?

**Comment by jbucaran** [comment]
@Swizz Is that not a problem also in your example?

**Comment by Swizz** [comment]
I am not highlighting problem, I am asking about your thought on it ^^

```jsx
 components: state => ({
    Counters: [
      <Counter initialValue={1} unit={state.unit} onchange={value => emit('change', {i: 1, value})}/>,
      <Counter initialValue={2} unit={state.unit} onchange={value => emit('change', {i: 2, value})}/>
      <Counter initialValue={3} unit={state.unit} onchange={value => emit('change', {i: 3, value})}/>
    ]
  }),
```

```jsx
 components: state => ({
    Counters: state.initialValues.map((value, i) =>
      <Counter initialValue={value} unit={state.unit} onchange={value => emit('change', {i, value})}/>
    )
  }),
```

With your 

```jsx
view: (state, actions, { Counter }, emit) =>
    <main>
      <h1>{state.title}</h1>
      <Counter initialValue={1} unit={state.unit} onchange={value => emit('change', {i: 1, value})}/>
      <Counter initialValue={2} unit={state.unit} onchange={value => emit('change', {i: 1, value})}/>
      <Counter initialValue={3} unit={state.unit} onchange={value => emit('change', {i: 1, value})}/>
    </main>,
```

```jsx
view: (state, actions, { Counter }, emit) =>
    <main>
      <h1>{state.title}</h1>
      {state.initialValues.map((value, i) =>
        <Counter initialValue={value} unit={state.unit} onchange={value => emit('change', {i, value})}/>
      )}
    </main>,
```

Both works the same, the main purpose is all about, the power we want to give to the view propertie.

My mind in stucks about separation and isolation ; your is on simplicity and intutivity.
Choose one. I am not responsible of that 😛

**Comment by jbucaran** [comment]
@Swizz What is unit by the way? And won't we have to provide an index to the component?

**Comment by Swizz** [comment]
Unit is just another attribute to illustrate parent->child communication and the index is only for the events example.

The id key could be set randomly or by given a key to the component like it does for vnode.

**Comment by jbucaran** [comment]
@Swizz What do you think of a random id?

**Comment by Swizz** [comment]
Not pure but easy to use and to think about

**Comment by jbucaran** [comment]
What would be a pure way to do it then?

**Comment by jbucaran** [comment]
@Swizz Just to be clear, we're not talking about keys as in keyed-vdom keys.

**Comment by Swizz** [comment]
The purpose could be the same "identify a thing". But the implementation is différente. The only purpose here is about isolation.

**Comment by naugtur** [comment]
God this is long. I'm in a different timezone apparently. I'm half way through and I need to comment on the image from @Swizz 

I really don't think the state should be built based on the parent-child relationships of components.

State is the most important part of the architecture and I (as an app developer) will never give up my decisions on the data strucutres used in state.

State should be structured the way developer wants and subtrees of state passed to components consciously. Anything else is a guaranteed disaster.

There are 3 sane ways to choose from, to pass a fragment of state into a component in order to scope it:
1. `component('key.in.state',...)`
1. `<Component scope={state.key.in.state}>`
1. 'connect' pattern from redux-react

Second option is the simplest, but it's based on a convention and might be easy to break.

Option 3 is implemented in my fork. https://github.com/naugtur/hyperapp/blob/master/poc/index.js

**Comment by zaceno** [comment]
Good point about state structure being sacred to the developer, @naugtur 

I'm not familiar with the connect-pattern from redux react, so let me check if I follow your poc correctly:

The "context" available to "connected" components, is defined in the `<Provider..` tag. And the provider doesn't output anything. It's just a function to define the context. And then any "connected" component will automatically have access to the context (without it needing to be passed as a prop), and other props work the same.

Is that correct?

That doesn't seem all that different from my proposal. The one difference I can see is that you have the provider defining what the context is, whereas I just bind the apps full state and actions to components. Do you agree or disagree?

Also like my proposal (well almost), it seems yours does not require changes in the app or h functions, and could be used via a separate package. True?

**Comment by jbucaran** [comment]
@naugtur In your example, the app has the actions that can up and down the counter. If that's going to be always the case, then what is the advantage of your controls/components from using custom tags?

**Comment by naugtur** [comment]
@jbucaran actually, I wouldn't want them to be much more than custom tags. Components (in my view) are custom tags that you don't have to pass all properties to, at least not explicitly.

**Comment by jbucaran** [comment]
@naugtur So they are custom tags that are bound to the state and actions?

**Comment by jbucaran** [comment]
Please have a look at @zaceno's [proposal](https://codepen.io/zaceno/pen/gRvWPx) and tell me if you two are on the same page?

**Comment by naugtur** [comment]
Yes. 

I would add limiting scope of state passed in, but overall it's what I'd like to use.

**Comment by jbucaran** [comment]
I hope I am not the only one having mixed feelings about components now. This discussion motivated me to revisit custom tags and mixins, as well as re-evaluate this project core values, but I am not even sure we have a problem anymore. I say, let's build one or two cool apps first and then revisit this issue.

To summarize my feelings about the issue, I'll quote @zaceno's:

> But the patterns I've seen so far have state and actions that are defined very independently. Even though we're merging their state and actions to the main tree, they don't seem to have any access to the main state tree. So how are the components meant to communicate? If the answer is: "via the props", then they are effectually no different to locally stateful components. And if that's the case, then what's the point of the "one-state-to-rule-them-all"?

Now, the good news is that some patterns have emerged:

### Widgets

@zaceno's state/action bound custom tags is interesting and could be published as an independent module. I like to think of them as "widgets". 🤓 

### Split-mixins

@MatejMazur's Split-mixins is a clever trick that requires no changes to core and lets you use mixins to achieve something similar to stateful components.

```js
const MyComponent = ({ state, view, actions }) => view
   ? ({ state, actions, events, mixins }) // good ol' mixin
    : <main>...</main> // a custom tag
```

## Final remarks

We're all (I am!) on a learning curve and tyring to figure this out. I am sure more patterns will show up, so please share them. I'll be opening a new issue to keep the discussion going.

Thank you all, for your feedback. This was an epic thread! 🎉 🥇

**Comment by MatejMazur** [comment]
[Here](https://github.com/MatejMazur/hyperapp-mixin) is package for mixins with view (it's just wrapper for my trick which @jbucaran mentioned in this [comment](https://github.com/hyperapp/hyperapp/issues/238#issuecomment-312455621). With this you can write mixins like this:

```jsx
const Counter = mixin({
  view: (state, actions, children) => (
    <div>
      <div>Number of counters: {state.length}</div>
      <button onclick={actions.addCounter}>Add counter</button>

      {state.map((instance, index) => (
        <div key={instance.id}>
          <button onclick={() => actions.increase(index)}>
            Increase
          </button>
          <span>{instance.sum}</span>
          <button onclick={() => actions.decrease(index)}>
            Decrease
          </button>
          <button onclick={() => actions.removeCounter(index)}>Remove</button>
        </div>
      ))}
    </div>
  ),

  state: {
    [COUNTER]: R.times(() => initial(), 2) // default number of loaded counters
  },

  actions: {
    [COUNTER]: {
      addCounter: (state) => ({
        [COUNTER]: R.append(initial(), state[COUNTER])
      }),
      removeCounter: (state, actions, index) => ({
        [COUNTER]: R.remove(index, 1, state[COUNTER])
      }),
      increase: (state, actions, index) => R.over(
        R.lensPath([COUNTER, index, 'sum']),
        R.inc,
        state
      ),
      decrease: (state, actions, index) => R.over(
        R.lensPath([COUNTER, index, 'sum']),
        R.dec,
        state
      )
    }
  }
})
```

and then it can be used in app like this:

```jsx
app({
  view: (state, actions) => (
    <Counter state={state.counter} actions={actions.counter} />
  ),
  ...,
  mixins: [Counter]
})
```
That's it!
